### ISSUE TITLE
用流水并行做训练时出现的问题： CrossEntropyLoss 报错及训练一定时间后阻塞

### ISSUE THREAD

**Issue by doombeaker** [Oneflow-Inc/oneflow#6224]
复现代码如下：

将以下代码存为 `test.py`，然后运行：

```shell
python -m oneflow.distributed.launch --nproc_per_node 2 ./test.py 
```

```python
import oneflow as flow
import oneflow.nn as nn
import oneflow.utils.vision.transforms as transforms

DEVICE = "cuda"
BROADCAST = [flow.sbp.broadcast]
P0 = flow.placement("cuda", {0: [0]})
P1 = flow.placement("cuda", {0: [1]})

mnist_train = flow.utils.vision.datasets.MNIST(
    root="data",
    train=True,
    transform=transforms.ToTensor(),
    download=True,
    source_url="https://oneflow-public.oss-cn-beijing.aliyuncs.com/datasets/mnist/MNIST/",
)
mnist_test = flow.utils.vision.datasets.MNIST(
    root="data",
    train=False,
    transform=transforms.ToTensor(),
    download=True,
    source_url="https://oneflow-public.oss-cn-beijing.aliyuncs.com/datasets/mnist/MNIST/",
)

train_iter = flow.utils.data.DataLoader(
    mnist_train, 16, shuffle=True
)
test_iter = flow.utils.data.DataLoader(
    mnist_test, 16, shuffle=False
)

class Stage0Module(flow.nn.Module):
    def __init__(self):
        super().__init__()
        self.flatten = flow.nn.Flatten()
        self.linear0 = flow.nn.Linear(28*28, 512)
        self.relu0 = flow.nn.ReLU()

        self.flatten.to_consistent(placement=P0, sbp=BROADCAST)
        self.linear0.to_consistent(placement=P0, sbp=BROADCAST)
        self.relu0.to_consistent(placement=P0, sbp=BROADCAST)

    def forward(self, x):
        out = self.flatten(x)
        out = self.linear0(out)
        out = self.relu0(out)
        return out

class Stage1Module(flow.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = flow.nn.Linear(512, 512)
        self.relu1 = flow.nn.ReLU()
        self.linear2 = flow.nn.Linear(512, 10)
        self.relu2 = flow.nn.ReLU()
        self.loss_fn = flow.nn.CrossEntropyLoss()
        
        self.linear1.to_consistent(placement=P1, sbp=BROADCAST)
        self.relu1.to_consistent(placement=P1, sbp=BROADCAST)
        self.linear2.to_consistent(placement=P1, sbp=BROADCAST)
        self.relu2.to_consistent(placement=P1, sbp=BROADCAST)
        self.loss_fn.to_consistent(placement=P1, sbp=BROADCAST)

    def forward(self, x, y):
        x = x.to_consistent(placement=P1, sbp=x.sbp)
        y = y.to_consistent(placement=P1, sbp=x.sbp)
        out = self.linear1(x)
        out = self.relu1(out)
        out = self.linear2(out)
        logits = self.relu2(out)
        print("yaochi:", logits.shape, y.shape)
        loss = self.loss_fn(logits, y)
        return loss

class PipelineModule(flow.nn.Module):
    def __init__(self):
        super().__init__()
        self.m_stage0 = Stage0Module()
        self.m_stage1 = Stage1Module()

    def forward(self, x, y):
        out_stage0 = self.m_stage0(x)
        out_stage1 = self.m_stage1(out_stage0, y)
        return out_stage1

module_pipeline = PipelineModule()
sgd = flow.optim.SGD(module_pipeline.parameters(), lr=0.0001)

class PipelineGraph(flow.nn.Graph):
    def __init__(self):
        super().__init__()
        self.module_pipeline = module_pipeline
        self.module_pipeline.m_stage0.config.stage_id = 0
        self.module_pipeline.m_stage1.config.stage_id = 1
        self.config.set_gradient_accumulation_steps(2)
        self.add_optimizer(sgd)

    def build(self, x, y):
        module_pipeline.train()
        loss = self.module_pipeline(x, y)
        loss.backward()
        return loss

graph_pipeline = PipelineGraph()

for x, y in train_iter:
    x = x.to_consistent(placement=P0, sbp=BROADCAST)
    y = y.to_consistent(placement=P0, sbp=BROADCAST)
    loss = graph_pipeline(x, y)
    if flow.env.get_rank() == 1:
        print(loss.to_local())
```

## 问题一

如果直接运行以上代码，会报错，称 **reshape 时，out shpae 和  in shape 不一致**：
```text
*** Check failure stack trace: ***
F0910 10:30:00.915789 2523627 op_graph.h:100] 
  File "/home/yaochi/oneflow/oneflow/core/graph/op_graph.h", line 100, in OpGraph
    Init(job)
  File "/home/yaochi/oneflow/oneflow/core/graph/op_graph.cpp", line 175, in Init
    InferLogicalBlobDesc(job)
  File "/home/yaochi/oneflow/oneflow/core/graph/op_graph.cpp", line 388, in InferLogicalBlobDesc
    TopoForEachNodeWithErrorCaptured([&](OpNode* op_node ... Data_YouAreNotAllowedToCallThisFuncOutsideThisFile(); return Maybe<void>::Ok(); })
  File "/home/yaochi/oneflow/oneflow/core/graph/graph.h", line 525, in TopoForEachNodeWithErrorCaptured
    Handler(cur_node)
  File "/home/yaochi/oneflow/oneflow/core/graph/op_graph.cpp", line 388, in operator()
    op_node->mut_op()->InferLogicalOutBlobDescsIf()
  File "/home/yaochi/oneflow/oneflow/core/operator/operator.cpp", line 308, in InferLogicalOutBlobDescsIf
    InferLogicalOutBlobDescs(BlobDesc4BnInOp, *({ auto&&  ... Data_YouAreNotAllowedToCallThisFuncOutsideThisFile())
  File "/home/yaochi/oneflow/oneflow/core/operator/user_op.cpp", line 568, in InferLogicalOutBlobDescs
    val_->logical_tensor_desc_infer_fn(&infer_ctx)
  File "/home/yaochi/oneflow/oneflow/user/ops/reshape_op.cpp", line 53, in LogicalTensorDescInferFn
    Check failed: (out_shape->elem_cnt()) == (in_shape.elem_cnt()) (160 vs 80)
```

经定位发现是 CrossEntropyLoss 里的这句报错：https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/nn/modules/loss.py#L186

如果把这句注释后，报错消失，可以训练一小段时间（但是会出现 **问题二**）

## 问题二
如果注释掉上文描述的的 CrossEntropyLoss 中的 reshape 代码，那么正常运行一段时间后，会报以下错误，然后阻塞：

```text
E0910 10:32:38.721472 2531154 blocking.cpp:28] [rank=0] Blocking detected. Python stack:
  File "./test.py", line 107, in <module>
    x = x.to_consistent(placement=P0, sbp=BROADCAST)
  File "/home/yaochi/oneflow/python/oneflow/nn/modules/consistent_cast.py", line 104, in to_consistent_op
    with BlockingInfoContext() as ctx:
  File "/home/yaochi/oneflow/python/oneflow/support/blocking.py", line 23, in __init__
    stack_info = "".join(traceback.format_stack(limit=5))
```

猜测第一个问题与流水并行中 micro batch 的引入，导致要调整 batch size 有关（但是我不知道如何具体操作）。
猜测第二个问题是未暴露的 bug。

**Comment by strint** [comment]
第二个会触发失败不，还是只是提示

**Comment by doombeaker** [comment]
> 第二个会触发失败不，还是只是提示

只是提示，然后卡死了

**Comment by Flowingsun007** [comment]
问题一，可能是demo写法的问题，应该不是reshape的问题（因为reshape要求input和output的元素个数一样：https://pytorch.org/docs/stable/generated/torch.reshape.html?highlight=reshape#torch.reshape）

**Comment by chengtbf** [comment]
这个是 reshape 的问题。  不是流水并行的问题，是我们 grad acc 的问题。 grad acc 会把 input unpack，batch 维的值就缩放了，后面的 reshape 就挂了。 这个问题我可以快速修一下。你这里可以先把 flatten 操作放到 Graph 外面可以绕过去  @doombeaker

**Comment by chengtbf** [comment]
这个脚本里有很多写的很糟糕的地方，我觉得不是一个合适的 demo：

### 1. Module 有 to_consistent 方法

Module 的 to_consistent 方法就是对内部的 self parameters 统一做 for each to_consistent 操作，所以下面的这个写法：


```
class Stage0Module(flow.nn.Module):
    def __init__(self):
        super().__init__()
        self.flatten = flow.nn.Flatten()
        self.linear0 = flow.nn.Linear(28*28, 512)
        self.relu0 = flow.nn.ReLU()

        self.flatten.to_consistent(placement=P0, sbp=BROADCAST)
        self.linear0.to_consistent(placement=P0, sbp=BROADCAST)
        self.relu0.to_consistent(placement=P0, sbp=BROADCAST)

    def forward(self, x):
        out = self.flatten(x)
        out = self.linear0(out)
        out = self.relu0(out)
        return out
```

可以改成：

```
class Stage0Module(flow.nn.Module):
    def __init__(self):
        super().__init__()
        self.flatten = flow.nn.Flatten()
        self.linear0 = flow.nn.Linear(28*28, 512)
        self.relu0 = flow.nn.ReLU()

    def forward(self, x):
        out = self.linear0(out)
        out = self.relu0(out)
        return out

my_module = Stage0Module()
my_module.to_consistent(placement=P0, sbp=BROADCAST)
```

**Comment by chengtbf** [comment]
### 2. stage 的 to_consistent 操作更新 Placement，不应该放在 Stage module 内部，而是放在外面的  PipelineModule 里：

```python
class Stage1Module(flow.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = flow.nn.Linear(512, 512)
        self.relu1 = flow.nn.ReLU()
        self.linear2 = flow.nn.Linear(512, 10)
        self.relu2 = flow.nn.ReLU()
        self.loss_fn = flow.nn.CrossEntropyLoss()

    def forward(self, x, y):
        x = x.to_consistent(placement=P1, sbp=x.sbp)
        y = y.to_consistent(placement=P1, sbp=x.sbp)
        out = self.linear1(x)
        out = self.relu1(out)
        out = self.linear2(out)
        logits = self.relu2(out)
        print("yaochi:", logits.shape, y.shape)
        loss = self.loss_fn(logits, y)
        return loss

```
不应该这么写，而是这样写：


```python
class Stage1Module(flow.nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, x, y):
        out = self.linear1(x)
        out = self.relu1(out)
        out = self.linear2(out)
        logits = self.relu2(out)
        print("yaochi:", logits.shape, y.shape)
        loss = self.loss_fn(logits, y)
        return loss

class PipelineModule(flow.nn.Module):
    def __init__(self):
        super().__init__()
        self.m_stage0 = Stage0Module()
        self.m_stage1 = Stage1Module()
        self.m_stage0.to_consistent(placement=P0, sbp=BROADCAST)
        self.m_stage1.to_consistent(placement=P0, sbp=BROADCAST)

    def forward(self, x, y):
        out_stage0 = self.m_stage0(x)
        in_stage1 = out_stage0.to_consistent(placement=P1, sbp=BROADCAST)
        y_stage1 = y.to_consistent(placement=P1, sbp=BROADCAST)
        out_stage1 = self.m_stage1(in_stage1 , y_stage1 )
        return out_stage1

```

to_consistent 操作应该是外面的胶水层 module 去关心的事情，而不是 Module 本身要关心的事情

**Comment by chengtbf** [comment]
### 3. 流水并行应该构造可以循环搭建 stage 的例子，而不是每个 stage 都特判：

流水并行里的中间 stage （除了第一个和最后一个），常常需要抽象好中间的 stage 对应的 module layer，主要的特征就是 layer 的输入和输出 shape 一致（比如 TransformerLayer），这样才能重复堆叠这些 stage。

```python

class StageLayerModule(flow.nn.Module):
    def __init__(self):
        super().__init__()
        self.in_linear = flow.nn.Linear(64, 10)
        self.out_linear = flow.nn.Linear(10, 64)
    def forward(self, x):
        # x.shape => (b, 64)
        h = self.in_linear(x) 
        # h.shape => (b, 10)
        h = F.relu(h)
        h = h * 2.0 + 5.0
        out = self.out_linear(h)
        out = F.relu(out)
        # x.shape => (b, 64)
        return out

for i in range(6):
    layer_list[i] = StageLayerModule()

# 在 pipeline module 里衔接每个中间的  layer
```

**Comment by doombeaker** [comment]
> 你这里可以先把 flatten 操作放到 Graph 外面可以绕过去 @doombeaker 而且 Flatten 要保持 Batch 维！

这个我没有懂，挂的地方并不是 flatten。是 CrossEntropyLoss 里的 reshape。为什么我把 flatten 放外面，就可以解决这个问题……

**Comment by chengtbf** [comment]
> 这个是 reshape 的问题。 不是流水并行的问题，是我们 grad acc 的问题。 grad acc 会把 input unpack，batch 维的值就缩放了，后面的 reshape 就挂了。 这个问题我可以快速修一下。
> 
> 你这里可以先把 flatten 操作放到 Graph 外面可以绕过去 @doombeaker 而且 Flatten 要保持 Batch 维！

哦哦哦和这个好像不是 flatten 的问题 。。。。 就是 reshape 在 grad acc 下 batch 维度缩放的问题，得直接改 reshape
